compat_unif

```diff
@@ -0,0 +1,144 @@
+Unification compatibility.
+
+TODO:
+
+  1 - Reorder commits to have changes to evarconv/clenv/clenvtac first.
+  2 - First apply changes
+  3 - Tactic changes
+  4 - rewrite changes.
+
+  5 - Rebase on current master.
+
+  6 - Copy old code in a plugin for compat and dispatch based on version.
+  
+  7 - Cleanup intf of new code, removing unification.ml and
+    metas (some thinking needed for patterns).
+
+  8 - Profit.
+
+Tactic refinements:
+
+  - apply should have two variants: one for triggering higher-order unification
+  and one without. apply_ho is much more powerful but can also produce meaningless
+  abstractions.
+
+
+- eapply could shelve subgoals before.
+
+  Two cases: forall e : Equiv, Reflexive (equiv e) applied to Reflexive ?X
+  -> Equiv becomes a subgoal even if dependent
+
+  ?X := equiv ?e
+
+  Proper (respectful (relprod ?R ?S) ?S) (snd t) applied to Proper (respectful ?R' eq) (@snd t t)
+
+  -> ?R does not become a subgoal, ?R dependent
+
+  No difference? ?MR is an initial _meta_ of the lemma, with ?R' :=
+  relprod ?R' ?S with ?MR := ?R', an evar.  ?Me is an initial meta of
+  the lemma with ?X := equiv ?e'.  ?Me := e'
+
+  Here ?R is a meta instantiated by an evar and unify_resolve succeeds
+  without seeing that evar is left around, as clenv_pose_dependent_evars
+  was just looking at the meta substitution which contains no meta
+  without instance.
```

Comment:
I wonder whether "eapply could shelve subgoals before" is related to this comment of [#15521](https://github.com/coq/coq/issues/15521#issuecomment-1416848455) about a difference between `apply` and `rapply`?